### PR TITLE
[VPlan] Model FOR extract of exit value in VPlan.

### PR DESCRIPTION
**Pull request by fhahn**
This patch introduces a new ExtractFromEnd VPInstruction opcode to extract the value of a FOR for users outside the loop (i.e. in the scalar loop's exits). This moves the first part of fixing first order recurrences to VPlan, and removes some additional code to patch up live-outs, which is now handled automatically.

The majority of test changes is due to changes in the order of which the extracts are generated now. As we are now using VPTransformState to generate the extracts, we may be able to re-use existing extracts in the loop body in some cases. For scalable vectors, in some cases we now have to compute the runtime VF twice, as each extract is now independent, but those should be trivial to clean up for later passes (and in line with other places in the code that also liberally re-compute runtime VFs).

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 1]
```diff
@@ -812,6 +812,8 @@ bool VPlanTransforms::adjustFixedOrderRecurrences(VPlan &Plan,
     if (auto *FOR = dyn_cast<VPFirstOrderRecurrencePHIRecipe>(&R))
       RecurrencePhis.push_back(FOR);
 
+  VPBuilder MiddleBuilder(
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:815]: Worth renaming `Builder` here to, say, `LoopBuilder`?

**Review by ayalz — COMMENTED** [review]
Nice step populating middle block with recipes!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -558,6 +560,27 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
 
     return ReducedPartRdx;
   }
+  case VPInstruction::ExtractRecurrenceResult: {
+    if (Part != 0)
+      return State.get(this, 0, /*IsScalar*/ true);
+
+    // Extract the second last element in the middle block for users outside the
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Comment is now gone after generalizing to ExtractFromEnd

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 3]
```diff
@@ -558,6 +560,27 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
 
     return ReducedPartRdx;
   }
+  case VPInstruction::ExtractRecurrenceResult: {
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Generalized to ExtractFromEnd and will update https://github.com/llvm/llvm-project/pull/93396 to use it to extract the resume value as well

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 4]
```diff
@@ -167,8 +167,8 @@ class VPLane {
 
   static VPLane getFirstLane() { return VPLane(0, VPLane::Kind::First); }
 
-  static VPLane getLastLaneForVF(const ElementCount &VF) {
-    unsigned LaneOffset = VF.getKnownMinValue() - 1;
+  static VPLane getLastLaneForVF(const ElementCount &VF, unsigned Offset = 1) {
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Generalized the patch to use ExtractFromEnd, added public `getLaneFromEnd` for that use case

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 5]
```diff
@@ -598,6 +621,7 @@ void VPInstruction::execute(VPTransformState &State) {
   bool GeneratesPerFirstLaneOnly =
       canGenerateScalarForFirstLane() &&
       (vputils::onlyFirstLaneUsed(this) ||
+       getOpcode() == VPInstruction::ExtractRecurrenceResult ||
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Can do, but reduction may not necessarily be a good fit for ExtractFromEnd. For now, they both generate a single scalar, either by reducing or extracting. How about `hasSingleScalarResult`?

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:815]: Updated, thanks!

File: llvm/test/Transforms/LoopVectorize/first-order-recurrence.ll [hunk 6]
```diff
@@ -1223,11 +1222,11 @@ define i32 @PR30183(i32 %pre_load, ptr %a, ptr %b, i64 %n) {
 ; SINK-AFTER-NEXT:    [[TMP14:%.*]] = getelementptr inbounds i32, ptr [[A]], i64 [[TMP10]]
 ; SINK-AFTER-NEXT:    [[TMP15:%.*]] = load i32, ptr [[TMP11]], align 4
 ; SINK-AFTER-NEXT:    [[TMP16:%.*]] = load i32, ptr [[TMP12]], align 4
-; SINK-AFTER-NEXT:    [[TMP17:%.*]] = load i32, ptr [[TMP13]], align 4
+; SINK-AFTER-NEXT:    [[VECTOR_RECUR_EXTRACT_FOR_PHI:%.*]] = load i32, ptr [[TMP13]], align 4
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/test/Transforms/LoopVectorize/first-order-recurrence.ll:1225]: Yes, this is a consequence of using State.get() to generate extracts, which will re-use existing extracts. (also mentioned in the commit message)

**Review by fhahn — COMMENTED** [review]
Updated and generalized to ExtracFromEnd opcode, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 7]
```diff
@@ -558,6 +560,27 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
...
+    Value *Res;
+    if (State.VF.isVector()) {
+      Res = State.get(
+          getOperand(0),
+          VPIteration(State.UF - 1, VPLane::getLastLaneForVF(State.VF, 2)));
+    } else {
+      assert(State.UF > 1 && "VF and UF cannot both be 1");
+      // When loop is unrolled without vectorizing, retrieve the value just
+      // prior to the final unrolled value. This is analogous to the vectorized
+      // case above: extracting the second last element when VF > 1.
```
[the remark this answers is not in the document]
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:580]: Comment still appears? Anyhow, "else" case corresponds to retrieving part UF - Offset "vertically".

**Review by fhahn — COMMENTED** [review]
Addressed latest comments, thanks!

Added new `isVectorToScalar` as suggested.

> Operator- would naturally return an ElementCount
An issue with that may be that we would need to create a VPLane from the element count first, to apply it's operator-, which may be almost as verbose as currently, which is a bit more explicit.

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 8]
```diff
@@ -167,8 +167,10 @@ class VPLane {
 
   static VPLane getFirstLane() { return VPLane(0, VPLane::Kind::First); }
 
-  static VPLane getLastLaneForVF(const ElementCount &VF) {
-    unsigned LaneOffset = VF.getKnownMinValue() - 1;
+  static VPLane getLaneFromEnd(const ElementCount &VF, unsigned Offset) {
+    assert(Offset <= VF.getKnownMinValue() &&
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlan.h:171]:
```suggestion
    assert(Offset > 0 && Offset <= VF.getKnownMinValue() &&
```

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 9]
```diff
@@ -831,18 +833,28 @@ bool VPlanTransforms::adjustFixedOrderRecurrences(VPlan &Plan,
...
     // Set the first operand of RecurSplice to FOR again, after replacing
     // all users.
     RecurSplice->setOperand(0, FOR);
+
+    Type *IntTy = Plan.getCanonicalIV()->getScalarType();
+    auto *Result = cast<VPInstruction>(MiddleBuilder.createNaryOp(
+        VPInstruction::ExtractFromEnd,
+        {FOR->getBackedgeValue(),
+         Plan.getOrAddLiveIn(ConstantInt::get(IntTy, 2))},
+        {}, "vector.recur.extract.for.phi"));
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:855]: ok, this is the unique Name assigned to this VPInstruction instance.

**Review by ayalz — APPROVED** [review]
Thanks for accommodating! Adding some final suggestions.

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 8 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Added, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 10]
```diff
@@ -1182,6 +1188,11 @@ class VPInstruction : public VPRecipeWithIRFlags {
     BranchOnCount,
     BranchOnCond,
     ComputeReductionResult,
+    // Takes the VPValue to extract from as first operand and the lane to
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Adjusted, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 11]
```diff
@@ -1182,6 +1188,11 @@ class VPInstruction : public VPRecipeWithIRFlags {
...
+    // extract from as second operand. The second operand must be a constant and
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Adjusted, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 12]
```diff
@@ -1182,6 +1188,11 @@ class VPInstruction : public VPRecipeWithIRFlags {
...
+    // <= VF when extracting from a vector or <= UF when extracting from a
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Adjusted, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 13]
```diff
@@ -1219,6 +1230,10 @@ class VPInstruction : public VPRecipeWithIRFlags {
   /// value for lane \p Lane.
   Value *generatePerLane(VPTransformState &State, const VPIteration &Lane);
 
+  /// Returns true if this VPInstruction converts a vector value to a scalar,
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Adjusted, thanks!

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 14]
```diff
@@ -3657,7 +3672,8 @@ inline bool isUniformAfterVectorization(VPValue *VPV) {
   if (auto *GEP = dyn_cast<VPWidenGEPRecipe>(Def))
     return all_of(GEP->operands(), isUniformAfterVectorization);
   if (auto *VPI = dyn_cast<VPInstruction>(Def))
-    return VPI->getOpcode() == VPInstruction::ComputeReductionResult;
+    return VPI->getOpcode() == VPInstruction::ComputeReductionResult ||
+           VPI->getOpcode() == VPInstruction::ExtractFromEnd;
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Updated, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 15]
```diff
@@ -137,6 +137,7 @@ bool VPRecipeBase::mayHaveSideEffects() const {
     case VPInstruction::Not:
     case VPInstruction::CalculateTripCountMinusVF:
     case VPInstruction::CanonicalIVIncrementForPart:
+    case VPInstruction::ExtractFromEnd:
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:140]: It is needed for the patch, as it always adds the extract and RAUW outside the loop, relying on DCE to clean up unneeded extracts.

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 16]
```diff
@@ -558,6 +559,29 @@ Value *VPInstruction::generatePerPart(VPTransformState &State, unsigned Part) {
 
     return ReducedPartRdx;
   }
+  case VPInstruction::ExtractFromEnd: {
+    if (Part != 0)
+      return State.get(this, 0, /*IsScalar*/ true);
+
+    auto *CI = cast<ConstantInt>(getOperand(1)->getLiveInIRValue());
+    unsigned Offset = CI->getZExtValue();
+
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Updated, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 9 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:855]: Yep

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:140]: Oh, this marks ExtractFromEnd as not having side effects.

**Pull request merged**